src/offline-support/offline_sync_manager.ts

```diff
@@ -0,0 +1,190 @@
+  };
+
+  /**
+   * Schedules a one-time callback to be invoked after the next successful sync.
+   *
+   * @param tag - A unique key to identify and manage the callback.
+   * @param callback - An async function to run after sync.
+   */
+  public scheduleSyncStatusChangeCallback = (
+    tag: string | symbol,
+    callback: () => Promise<void>,
+  ) => {
+    this.scheduledSyncStatusCallbacks.set(tag, callback);
+  };
+
+  /**
+   * Invokes all registered sync status listeners and executes any scheduled sync callbacks.
+   *
+   * @param status - The new sync status (`true` or `false`).
+   */
+  private invokeSyncStatusListeners = async (status: boolean) => {
+    this.syncStatus = status;
+    this.syncStatusListeners.forEach((l) => l(status));
+
+    if (status) {
+      const promises = Array.from(this.scheduledSyncStatusCallbacks.values()).map((cb) =>
+        cb(),
+      );
+      await Promise.all(promises);
+
+      this.scheduledSyncStatusCallbacks.clear();
+    }
+  };
+
+  /**
+   * Performs synchronization with the Stream backend.
+   *
+   * This includes downloading events since the last sync, updating the local DB,
+   * and handling sync failures (e.g., if syncing beyond the allowed retention window).
+   */
+  private sync = async () => {
+    if (!this.client?.user) {
+      return;
+    }
+    const cids = await this.offlineDb.getAllChannelCids();
+    // If there are no channels, then there is no need to sync.
+    if (cids.length === 0) {
+      return;
+    }
+
+    // TODO: We should not need our own user ID in the API, it can be inferred
+    const lastSyncedAt = await this.offlineDb.getLastSyncedAt({
+      userId: this.client.user.id,
+    });
+
+    if (lastSyncedAt) {
+      const lastSyncedAtDate = new Date(lastSyncedAt);
+      const nowDate = new Date();
+
+      // Calculate the difference in days
+      const diff = Math.floor(
+        (nowDate.getTime() - lastSyncedAtDate.getTime()) / (1000 * 60 * 60 * 24),
+      );
+
+      if (diff > 30) {
+        // stream backend will send an error if we try to sync after 30 days.
+        // In that case reset the entire DB and start fresh.
+        await this.offlineDb.resetDB();
+      } else {
+        try {
+          const result = await this.client.sync(cids, lastSyncedAtDate.toISOString());
+          const queryPromises = result.events.map((event) =>
+            this.offlineDb.handleEvent({ event, execute: false }),
+          );
+          const queriesArray = await Promise.all(queryPromises);
+          const queries = queriesArray.flat() as ExecuteBatchQueriesType;
+
+          if (queries.length) {
+            await this.offlineDb.executeSqlBatch(queries);
+          }
+        } catch (e) {
+          console.log('An error has occurred while syncing the DB.', e);
+          // Error will be raised by the sync API if there are too many events.
+          // In that case reset the entire DB and start fresh.
+          await this.offlineDb.resetDB();
```

Comment:
Maybe re-throwing an error here might sound better - let the user of the API handle it themselves.

Reply:
(useful for some UI stuff - like toasts and whatnot)

Reply:
That's unfortunate, consider resolved then. 👍